File: org.knime.python2/src/org/knime/python2/util/PythonUtils.java
/*
 * ------------------------------------------------------------------------
 *
 *  Copyright by KNIME AG, Zurich, Switzerland
 *  Website: http://www.knime.com; Email: contact@knime.com
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License, Version 3, as
 *  published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, see <http://www.gnu.org/licenses>.
 *
 *  Additional permission under GNU GPL version 3 section 7:
 *
 *  KNIME interoperates with ECLIPSE solely via ECLIPSE's plug-in APIs.
 *  Hence, KNIME and ECLIPSE are both independent programs and are not
 *  derived from each other. Should, however, the interpretation of the
 *  GNU GPL Version 3 ("License") under any applicable laws result in
 *  KNIME and ECLIPSE being a combined program, KNIME AG herewith grants
 *  you the additional permission to use and propagate KNIME together with
 *  ECLIPSE with only the license terms in place for ECLIPSE applying to
 *  ECLIPSE and the GNU GPL Version 3 applying for KNIME, provided the
 *  license terms of ECLIPSE themselves allow for the respective use and
 *  propagation of ECLIPSE together with KNIME.
 *
 *  Additional permission relating to nodes for KNIME that extend the Node
 *  Extension (and in particular that are based on subclasses of NodeModel,
 *  NodeDialog, and NodeView) and that only interoperate with KNIME through
 *  standard APIs ("Nodes"):
 *  Nodes are deemed to be separate and independent programs and to not be
 *  covered works.  Notwithstanding anything to the contrary in the
 *  License, the License does not apply to Nodes, you are not required to
 *  license Nodes under the License, and you are granted a license to
 *  prepare and propagate Nodes, in each case even if such Nodes are
 *  propagated with or for interoperation with KNIME.  The owner of a Node
 *  may freely choose the license terms applicable to such Node, including
 *  when such Node is propagated with or for interoperation with KNIME.
 * ---------------------------------------------------------------------
 *
 * History
 *   May 24, 2018 (marcel): created
 */
package org.knime.python2.util;

import java.util.Arrays;
import java.util.Collection;
import java.util.Optional;
import java.util.concurrent.Callable;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Function;

import org.knime.core.node.CanceledExecutionException;
import org.knime.python2.PythonFrameSummary;
import org.knime.python2.kernel.PythonCancelable;
import org.knime.python2.kernel.PythonCanceledExecutionException;
import org.knime.python2.kernel.PythonException;
import org.knime.python2.kernel.PythonIOException;

/**
 * @author Marcel Wiedenmann, KNIME GmbH, Konstanz, Germany
 * @author Christian Dietz, KNIME GmbH, Konstanz, Germany
 */
public final class PythonUtils {

    private PythonUtils() {
        // utility class
    }

    /**
     * Utility class that helps checking whether the preconditions of a method or constructor invocation have been met
     * by the caller. <br>
     * This class complements the functionality of {@link com.google.common.base.Preconditions}.
     */
    public static class Preconditions {

        private Preconditions() {}

        /**
         * Ensures that a string passed as a parameter to the calling method is not null or empty.
         *
         * @param string a string
         * @return the non-null and non-empty reference that was validated
         * @throws NullPointerException if the input is null
         * @throws IllegalArgumentException if the input is empty
         * @see com.google.common.base.Preconditions#checkNotNull(Object)
         */
        public static String checkNotNullOrEmpty(final String string) {
            if (string == null) {
                throw new NullPointerException();
            }
            if (string.isEmpty()) {
                throw new IllegalArgumentException();
            }
            return string;
        }

        /**
         * Ensures that a string passed as a parameter to the calling method is not null or empty.
         *
         * @param string a string
         * @param errorMessage the exception message to use if the check fails; will be converted to a string using
         *            {@link String#valueOf(Object)}
         * @return the non-null and non-empty reference that was validated
         * @throws NullPointerException if the input is null
         * @throws IllegalArgumentException if the input is empty
         * @see com.google.common.base.Preconditions#checkNotNull(Object, Object)
         */
        public static String checkNotNullOrEmpty(final String string, final Object errorMessage) {
            if (string == null) {
                throw new NullPointerException(String.valueOf(errorMessage));
            }
            if (string.isEmpty()) {
                throw new IllegalArgumentException(String.valueOf(errorMessage));
            }
            return string;
        }

        /**
         * Ensures that a collection passed as a parameter to the calling method is not null or empty.
         *
         * @param collection a collection
         * @return the non-null and non-empty reference that was validated
         * @throws NullPointerException if the input is null
         * @throws IllegalArgumentException if the input is empty
         * @see com.google.common.base.Preconditions#checkNotNull(Object)
         */
        public static <C extends Collection<?>> C checkNotNullOrEmpty(final C collection) {
            if (collection == null) {
                throw new NullPointerException();
            }
            if (collection.isEmpty()) {
                throw new IllegalArgumentException();
            }
            return collection;
        }

        /**
         * Ensures that a collection passed as a parameter to the calling method is not null or empty.
         *
         * @param collection a collection
         * @param errorMessage the exception message to use if the check fails; will be converted to a string using
         *            {@link String#valueOf(Object)}
         * @return the non-null and non-empty reference that was validated
         * @throws NullPointerException if the input is null
         * @throws IllegalArgumentException if the input is empty
         * @see com.google.common.base.Preconditions#checkNotNull(Object, Object)
         */
        public static <C extends Collection<?>> C checkNotNullOrEmpty(final C collection, final Object errorMessage) {
            if (collection == null) {
                throw new NullPointerException(String.valueOf(errorMessage));
            }
            if (collection.isEmpty()) {
                throw new IllegalArgumentException(String.valueOf(errorMessage));
            }
            return collection;
        }
    }

    public static class Misc {

        private Misc() {}

        /**
         * @param exceptionConsumer may be <code>null</code>. If non-<code>null</code>, is used to report exceptions
         *            that occur while closing the individual given closeables.
         * @param closeables the resources to {@link AutoCloseable#close() close}. May be <code>null</code>, individual
         *            closeables may be <code>null</code>.
         * @throws Error if any occurred. If multiple {@link Error errors} occurred, the first one is thrown and the
         *             others are added as {@link Error#addSuppressed(Throwable) suppressed} if possible.<br>
         *             Note that this method never throws {@link Exception exceptions}.
         */
        public static void closeSafelyThrowErrors(final BiConsumer<String, Exception> exceptionConsumer,
            final AutoCloseable... closeables) {
            final Error error = closeSafely(exceptionConsumer, closeables);
            if (error != null) {
                throw error;
            }
        }

        /**
         * @param exceptionConsumer may be <code>null</code>. If non-<code>null</code>, is used to report exceptions
         *            that occur while closing the individual given closeables.
         * @param closeables the resources to {@link AutoCloseable#close() close}. May be <code>null</code>, individual
         *            closeables may be <code>null</code>.
         * @throws Error if any occurred. If multiple {@link Error errors} occurred, the first one is thrown and the
         *             others are added as {@link Error#addSuppressed(Throwable) suppressed} if possible.<br>
         *             Note that this method never throws {@link Exception exceptions}.
         */
        public static void closeSafelyThrowErrors(final BiConsumer<String, Exception> exceptionConsumer,
            final Iterable<? extends AutoCloseable> closeables) {
            final Error error = closeSafely(exceptionConsumer, closeables);
            if (error != null) {
                throw error;
            }
        }

        /**
         * @param exceptionConsumer may be <code>null</code>. If non-<code>null</code>, is used to report exceptions
         *            that occur while closing the individual given closeables.
         * @param closeables the resources to {@link AutoCloseable#close() close}. May be <code>null</code>, individual
         *            closeables may be <code>null</code>.
         * @return an error if any occurred, <code>null</code> otherwise. If non-<code>null</code>, it is recommended to
         *         rethrow that error. If multiple errors occurred, the first one is returned and the others are added
         *         as {@link Error#addSuppressed(Throwable) suppressed} if possible.
         *
         */
        public static Error closeSafely(final BiConsumer<String, Exception> exceptionConsumer,
            final AutoCloseable... closeables) {
            if (closeables == null) {
                return null;
            }
            return closeSafely(exceptionConsumer, Arrays.asList(closeables));
        }

        /**
         * @param exceptionConsumer may be <code>null</code>. If non-<code>null</code>, is used to report exceptions
         *            that occur while closing the individual given closeables.
         * @param closeables the resources to {@link AutoCloseable#close() close}. May be <code>null</code>, individual
         *            closeables may be <code>null</code>.
         * @return an error if any occurred, <code>null</code> otherwise. If non-<code>null</code>, it is recommended to
         *         rethrow that error. If multiple errors occurred, the first one is returned and the others are added
         *         as {@link Error#addSuppressed(Throwable) suppressed} if possible.
         *
         */
        public static Error closeSafely(final BiConsumer<String, Exception> exceptionConsumer,
            final Iterable<? extends AutoCloseable> closeables) {
            return invokeSafely(exceptionConsumer, closeable -> {
                try {
                    closeable.close();
                } catch (final Exception ex) {
                    throw new RuntimeException(ex.getMessage(), ex);
                }
            }, closeables);
        }

        /**
         * Submits the given task to the given executor service and blocks until the task is finished, an exception
         * occurs, or the execution is canceled. The task is {@link Thread#interrupt() interrupted} if a cancellation
         * occurs.
         *
         * @param task the task to run cancelable
         * @param executorService the executor service to which to submit the task
         * @param cancelable the cancelable to check for cancellation
         * @return the result of the task, if any
         * @throws RejectedExecutionException if the task cannot be submitted to the executor service
         * @throws PythonIOException if any exception occurred during execution (except cancellation, see below)
         * @throws PythonCanceledExecutionException if canceled, is also thrown if the task itself terminates due to
         *             {@link PythonCanceledExecutionException}, {@link CanceledExecutionException}, or
         *             {@link CancellationException}
         */
        public static <T> T executeCancelable(final Callable<T> task, final ExecutorService executorService,
            final PythonCancelable cancelable) throws PythonIOException, PythonCanceledExecutionException {
            final Future<T> future = executorService.submit(task);
            // Wait until execution is done or cancelled.
            final int waitTimeoutMilliseconds = 1000;
            while (true) {
                try {
                    return future.get(waitTimeoutMilliseconds, TimeUnit.MILLISECONDS);
                } catch (TimeoutException | InterruptedException ex) {
                    // The current thread has been interrupted or the task is not done yet.
                    if (ex instanceof InterruptedException) {
                        Thread.currentThread().interrupt();
                    }
                    try {
                        cancelable.checkCanceled();
                    } catch (final PythonCanceledExecutionException cancellation) {
                        // Execution was canceled, cancel task.
                        future.cancel(true);
                        throw cancellation;
                    }
                } catch (final CancellationException ex) {
                    // Should not happen since the handle to the future is local to this method.
                    throw new PythonCanceledExecutionException();
                } catch (final ExecutionException wrapper) {
                    final Throwable ex = unwrapExecutionException(wrapper).orElse(wrapper);
                    if (ex instanceof PythonCanceledExecutionException) {
                        // May happen if the executed task checks for cancellation itself.
                        throw (PythonCanceledExecutionException)ex;
                    } else if (ex instanceof CanceledExecutionException || ex instanceof CancellationException) {
                        // May happen if the executed task checks for cancellation itself.
                        throw new PythonCanceledExecutionException(ex.getMessage());
                    } else {
                        throw new PythonIOException(ex.getMessage(), ex);
                    }
                }
            }
        }

        /**
         * @param exceptionConsumer may be <code>null</code>. If non-<code>null</code>, is used to report exceptions
         *            that occur while invoking the given method on the individual given invokees.
         * @param method the method to invoke, may be <code>null</code>
         * @param invokees the objects on which to invoke the given method. May be <code>null</code>, individual objects
         *            may be <code>null</code>.
         * @return an error if any occurred, <code>null</code> otherwise. If non-<code>null</code>, it is recommended to
         *         rethrow that error. If multiple errors occurred, the first one is returned and the others are added
         *         as {@link Error#addSuppressed(Throwable) suppressed} if possible.
         */
        @SafeVarargs
        public static <T> Error invokeSafely(final BiConsumer<String, Exception> exceptionConsumer,
            final Consumer<? super T> method, final T... invokees) {
            if (invokees == null) {
                return null;
            }
            return invokeSafely(exceptionConsumer, method, Arrays.asList(invokees));
        }

        /**
         * @param exceptionConsumer may be <code>null</code>. If non-<code>null</code>, is used to report exceptions
         *            that occur while invoking the given method on the individual given invokees.
         * @param method the method to invoke, may be <code>null</code>
         * @param invokees the objects on which to invoke the given method. May be <code>null</code>, individual objects
         *            may be <code>null</code>.
         * @return an error if any occurred, <code>null</code> otherwise. If non-<code>null</code>, it is recommended to
         *         rethrow that error. If multiple errors occurred, the first one is returned and the others are added
         *         as {@link Error#addSuppressed(Throwable) suppressed} if possible.
         */
        public static <T> Error invokeSafely(final BiConsumer<String, Exception> exceptionConsumer,
            final Consumer<? super T> method, final Iterable<? extends T> invokees) {
            if (method == null || invokees == null) {
                return null;
            }
            Error error = null;
            for (final T i : invokees) {
                try {
                    if (i != null) {
                        method.accept(i);
                    }
                } catch (final Exception ex) {
                    if (exceptionConsumer != null) {
                        try {
                            exceptionConsumer.accept("An exception occurred while safely invoking a method."
                                + (ex.getMessage() != null ? " Exception: " + ex.getMessage() : ""), ex);
                        } catch (final Exception ignore) {
                            // ignore
                        } catch (final Error e) {
                            // Error will be handled by caller.
                            if (error == null) {
                                error = e;
                            } else {
                                error.addSuppressed(e);
                            }
                        }
                    }
                } catch (final Error e) {
                    // Error will be handled by caller.
                    if (error == null) {
                        error = e;
                    } else {
                        error.addSuppressed(e);
                    }
                }
            }
            return error;
        }

        /**
         * Finds the topmost throwable in a {@link Throwable#getCause() stack} of throwables that is not an
         * {@link ExecutionException}.
         *
         * @param throwable the topmost throwable in the stack of throwables
         * @return an optional that contains the throwable if one is found
         */
        public static Optional<Throwable> unwrapExecutionException(final Throwable throwable) {
            return traverseStackUntilFound(throwable, t -> t instanceof ExecutionException ? null : t);
        }

        /**
         * Finds the topmost {@link PythonException#getPythonTraceback() Python traceback} in a
         * {@link Throwable#getCause() stack} of throwables.
         *
         * @param throwable the topmost throwable in the stack of throwables
         * @return an optional that contains the traceback if one is found
         */
        public static Optional<PythonFrameSummary[]> extractPythonTraceback(final Throwable throwable) {
            return traverseStackUntilFound(throwable, t -> {
                if (t instanceof PythonException) {
                    return ((PythonException)t).getPythonTraceback().orElse(null);
                } else {
                    return null;
                }
            });
        }

        /**
         * Finds the topmost {@link PythonException#getShortMessage() Python short message} in a
         * {@link Throwable#getCause() stack} of throwables.
         *
         * @param throwable the topmost throwable in the stack of throwables
         * @return an optional that contains the short message if one is found
         */
        public static Optional<String> extractPythonShortMessage(final Throwable throwable) {
            return traverseStackUntilFound(throwable, t -> {
                if (t instanceof PythonException) {
                    return ((PythonException)t).getShortMessage().orElse(null);
                } else {
                    return null;
                }
            });
        }

        /**
         * Finds the topmost {@link PythonException#getFormattedPythonTraceback() formatted Python traceback} in a
         * {@link Throwable#getCause() stack} of throwables.
         *
         * @param throwable the topmost throwable in the stack of throwables
         * @return an optional that contains the formatted traceback if one is found
         */
        public static Optional<String> extractFormattedPythonTraceback(final Throwable throwable) {
            return traverseStackUntilFound(throwable, t -> {
                if (t instanceof PythonException) {
                    return ((PythonException)t).getFormattedPythonTraceback().orElse(null);
                } else {
                    return null;
                }
            });
        }

        private static <T> Optional<T> traverseStackUntilFound(final Throwable throwable,
            final Function<Throwable, T> visitor) {
            if (throwable == null) {
                return Optional.empty();
            }
            Throwable t = throwable;
            do {
                final T visitOutcome = visitor.apply(t);
                if (visitOutcome != null) {
                    return Optional.of(visitOutcome);
                }
            } while (t != t.getCause() && (t = t.getCause()) != null);
            return Optional.empty();
        }
    }
}

File: org.knime.python2/py/PythonUtils.py
# -*- coding: utf-8 -*-
# ------------------------------------------------------------------------
#  Copyright by KNIME AG, Zurich, Switzerland
#  Website: http://www.knime.com; Email: contact@knime.com
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License, Version 3, as
#  published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful, but
#  WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, see <http://www.gnu.org/licenses>.
#
#  Additional permission under GNU GPL version 3 section 7:
#
#  KNIME interoperates with ECLIPSE solely via ECLIPSE's plug-in APIs.
#  Hence, KNIME and ECLIPSE are both independent programs and are not
#  derived from each other. Should, however, the interpretation of the
#  GNU GPL Version 3 ("License") under any applicable laws result in
#  KNIME and ECLIPSE being a combined program, KNIME AG herewith grants
#  you the additional permission to use and propagate KNIME together with
#  ECLIPSE with only the license terms in place for ECLIPSE applying to
#  ECLIPSE and the GNU GPL Version 3 applying for KNIME, provided the
#  license terms of ECLIPSE themselves allow for the respective use and
#  propagation of ECLIPSE together with KNIME.
#
#  Additional permission relating to nodes for KNIME that extend the Node
#  Extension (and in particular that are based on subclasses of NodeModel,
#  NodeDialog, and NodeView) and that only interoperate with KNIME through
#  standard APIs ("Nodes"):
#  Nodes are deemed to be separate and independent programs and to not be
#  covered works.  Notwithstanding anything to the contrary in the
#  License, the License does not apply to Nodes, you are not required to
#  license Nodes under the License, and you are granted a license to
#  prepare and propagate Nodes, in each case even if such Nodes are
#  propagated with or for interoperation with KNIME.  The owner of a Node
#  may freely choose the license terms applicable to such Node, including
#  when such Node is propagated with or for interoperation with KNIME.
# ------------------------------------------------------------------------

"""
@author Clemens von Schwerin, KNIME GmbH, Konstanz, Germany
@author Patrick Winter, KNIME GmbH, Konstanz, Germany
@author Marcel Wiedenmann, KNIME GmbH, Konstanz, Germany
@author Christian Dietz, KNIME GmbH, Konstanz, Germany
"""

# This should be the first statement in each module that makes specific demands on the Python environment.
import EnvironmentHelper

if EnvironmentHelper.is_python3():
    import importlib
else:
    import imp

if EnvironmentHelper.is_tslib_available():
    from EnvironmentHelper import Timestamp
    from EnvironmentHelper import NaT
else:
    Timestamp = None
    NaT = None

import base64
import inspect
import math
import numpy
import os

from datetime import datetime

# list of equivalent types
EQUIVALENT_TYPES = []

# fill EQUIVALENT_TYPES
if not EnvironmentHelper.is_python3():
    EQUIVALENT_TYPES.append([unicode, str])  # Available in Python 2.
    EQUIVALENT_TYPES.append([int, long])  # Available in Python 2.

if EnvironmentHelper.is_tslib_available():
    EQUIVALENT_TYPES.append([datetime, Timestamp])

# Do not change those to sets. We need __eq__ comparison (overridden by numpy.dtypes), not __hash__.
# See https://docs.scipy.org/doc/numpy-1.14.2/reference/arrays.scalars.html.
_BOOLEAN_TYPES = ('bool', 'bool_', 'bool8')

_DOUBLE_TYPES = ('float', 'float_', 'double',
                 'float64')

_FLOAT_TYPES = ('half', 'single', 'float16', 'float32')

_INTEGER_TYPES = ('byte', 'short', 'int',
                  'intc', 'int_', 'longlong',
                  'intp', 'int8', 'int16',
                  'int32', 'int64', 'ubyte',
                  'ushort', 'uintc', 'uint',
                  'ulonglong', 'uintp', 'uint8',
                  'uint16', 'uint32', 'uint64')


class Simpletype:
    """
    Enum containing ids for all simple types, i.e. types that can be serialized directly using the serialization
    library.
    """
    BOOLEAN = 1
    BOOLEAN_LIST = 2
    BOOLEAN_SET = 3
    INTEGER = 4
    INTEGER_LIST = 5
    INTEGER_SET = 6
    LONG = 7
    LONG_LIST = 8
    LONG_SET = 9
    DOUBLE = 10  # np.float64 and Python floats
    DOUBLE_LIST = 11
    DOUBLE_SET = 12
    STRING = 13
    STRING_LIST = 14
    STRING_SET = 15
    BYTES = 16
    BYTES_LIST = 17
    BYTES_SET = 18
    FLOAT = 19  # np.float32 and smaller numpy floats
    FLOAT_LIST = 20
    FLOAT_SET = 21


def is_collection(data_type):
    """
    Checks if the given type is a collection type.
    """
    return (data_type is list  #
            or data_type is tuple  #
            or data_type is set  #
            or data_type is frozenset  #
            or data_type is dict)


def is_missing(value):
    """
    Checks if the given value is None, NaN or NaT.
    """
    return value is None or is_nat(value) or is_nan(value)


def is_nan(value):
    """
    Checks if the given value is NaN.
    """
    try:
        return math.isnan(value)
    except BaseException:
        return False


def is_nat(value):
    """
    Checks if the given value is NaT.
    """
    if EnvironmentHelper.is_tslib_available():
        return value is NaT
    else:
        return False


def is_numpy_type(data_type):
    """
    Checks if the given type is a numpy type, or, if it's an instance of a class, checks if its class is a numpy type.
    """
    if not inspect.isclass(data_type):
        data_type = type(data_type)
    return data_type.__module__ == numpy.__name__


def is_boolean_type(data_type):
    return data_type in _BOOLEAN_TYPES


def is_double_type(data_type):
    return data_type in _DOUBLE_TYPES


def is_float_type(data_type):
    return data_type in _FLOAT_TYPES


def is_integer_type(data_type):
    """
    Note, "integer" is to be understood in its mathematical sense, i.e., as a number without a fractional component.
    This does include, but is _not_ equivalent to, the unsigned 32-bit numeric data type.
    """
    return data_type in _INTEGER_TYPES


def fits_into_java_integer(data_type):
    """
    :param data_type: It is assumed that is_integer_type returns True for this argument.
    """
    return data_type.itemsize < 4 or (data_type.itemsize == 4 and data_type.kind == 'i')


def fits_into_java_long(data_type):
    """
    :param data_type: It is assumed that is_integer_type returns True for this argument.
    """
    return data_type.itemsize < 8 or (data_type.itemsize == 8 and data_type.kind == 'i')


def is_unsigned_byte_type(data_type):
    """
    "unsigned byte" refers to all flavors of the unsigned numeric data type of 8 bits of size.
    """
    return numpy.dtype('uint8') == data_type


def types_are_equivalent(type_1, type_2):
    """
    Checks if the two given types are equivalent based on the equivalence list and the equivalence of numpy types to
    python types.
    """
    for pair in EQUIVALENT_TYPES:
        if type_1 is pair[0] and type_2 is pair[1]:
            return True
        if type_1 is pair[1] and type_2 is pair[0]:
            return True
    if is_collection(type_1) or is_collection(type_2):
        return type_1 is type_2
    if is_numpy_type(type_1) or is_numpy_type(type_2):
        return numpy.issubdtype(type_1, type_2) and numpy.issubdtype(type_2, type_1)
    else:
        return type_1 is type_2


def get_type_string(data_object):
    """
    Gets the name of an object's type.
    NOTE: The name of an object's type is not the fully qualified one returned by the type() function but just the last
    piece of it (e.g. 'time' instead of 'datetime.time').
    """
    if hasattr(data_object, '__module__'):
        return data_object.__module__ + '.' + data_object.__class__.__name__
    else:
        return data_object.__class__.__name__


def object_to_string(data_object):
    """
    Convert data_object to a (possibly truncated) string representation.
    """
    if EnvironmentHelper.is_python3():
        try:
            object_as_string = str(data_object)
        except Exception:
            return ''
    else:
        try:
            object_as_string = unicode(data_object)
        except UnicodeDecodeError:
            object_as_string = '(base64 encoded)\n' + base64.b64encode(data_object)
        except Exception:
            return ''
    return (object_as_string[:996] + '\n...') if len(object_as_string) > 1000 else object_as_string


def value_to_simpletype_value(value, simpletype):
    """
    Convert a value into a given {@link Simpletype}.
    @param value the value to convert
    @param simpletype the {@link Simpletype} to convert into
    """
    if value is None:
        return None
    elif simpletype == Simpletype.BOOLEAN:
        return bool(value)
    elif simpletype == Simpletype.BOOLEAN_LIST:
        for i in range(0, len(value)):
            if value[i] is not None:
                value[i] = bool(value[i])
        return value
    elif simpletype == Simpletype.BOOLEAN_SET:
        value_set = set()
        for inner_value in value:
            if inner_value is None:
                value_set.add(None)
            else:
                value_set.add(bool(inner_value))
        return value_set
    elif simpletype == Simpletype.INTEGER:
        return int(value)
    elif simpletype == Simpletype.INTEGER_LIST:
        for i in range(0, len(value)):
            if value[i] is not None:
                value[i] = int(value[i])
        return value
    elif simpletype == Simpletype.INTEGER_SET:
        value_set = set()
        for inner_value in value:
            if inner_value is None:
                value_set.add(None)
            else:
                value_set.add(int(inner_value))
        return value_set
    elif simpletype == Simpletype.LONG:
        return int(value)
    elif simpletype == Simpletype.LONG_LIST:
        for i in range(0, len(value)):
            if value[i] is not None:
                value[i] = int(value[i])
        return value
    elif simpletype == Simpletype.LONG_SET:
        value_set = set()
        for inner_value in value:
            if inner_value is None:
                value_set.add(None)
            else:
                value_set.add(int(inner_value))
        return value_set
    elif simpletype == Simpletype.DOUBLE or simpletype == Simpletype.FLOAT:
        float_value = float(value)
        return float_value
    elif simpletype == Simpletype.DOUBLE_LIST or simpletype == Simpletype.FLOAT_LIST:
        for i in range(0, len(value)):
            if value[i] is not None:
                float_value = float(value[i])
                value[i] = float_value
        return value
    elif simpletype == Simpletype.DOUBLE_SET or simpletype == Simpletype.FLOAT_SET:
        value_set = set()
        for inner_value in value:
            if inner_value is None:
                value_set.add(None)
            else:
                float_value = float(inner_value)
                value_set.add(float_value)
        return value_set
    elif simpletype == Simpletype.STRING:
        return str(value)
    elif simpletype == Simpletype.STRING_LIST:
        for i in range(0, len(value)):
            if value[i] is not None:
                value[i] = str(value[i])
        return value
    elif simpletype == Simpletype.STRING_SET:
        value_set = set()
        for inner_value in value:
            if inner_value is None:
                value_set.add(None)
            else:
                value_set.add(str(inner_value))
        return value_set
    elif simpletype == Simpletype.BYTES:
        return bytes(value)
    elif simpletype == Simpletype.BYTES_LIST:
        for i in range(0, len(value)):
            if value[i] is not None:
                value[i] = bytes(value[i])
        return value
    elif simpletype == Simpletype.BYTES_SET:
        value_set = set()
        for inner_value in value:
            if inner_value is None:
                value_set.add(None)
            else:
                value_set.add(bytes(inner_value))
        return value_set


def bytes_from_file(path):
    return open(path, 'rb').read()


def load_module_from_path(path):
    """
    Load a python module from a source file.
    @param path the path to the source file (string)
    @return the module loaded from the specified path
    """
    last_separator = path.rfind(os.sep)
    file_extension_start = path.rfind('.')
    module_name = path[last_separator + 1:file_extension_start]
    try:
        if EnvironmentHelper.is_python3():
            loaded_module = importlib.machinery.SourceFileLoader(module_name, path).load_module()
        else:
            loaded_module = imp.load_source(module_name, path)
    except ImportError as error:
        raise ImportError('Error while loading python module ' + module_name + '\nCause: ' + str(error))
    return loaded_module


def invoke_safely(exception_consumer, method, invokees):
    if method is None or invokees is None:
        return None
    if not isinstance(invokees, list):
        invokees = [invokees]
    error = None
    for i in invokees:
        try:
            if i is not None:
                method(i)
        except Exception as ex:
            if exception_consumer is not None:
                try:
                    exception_consumer("An exception occurred while safely invoking a method. Exception: " + str(ex),
                                       ex)
                except Exception:
                    pass  # ignore
                except BaseException as ex:
                    # Error will be handled by caller.
                    if error is None:
                        error = ex
        except BaseException as ex:
            # Error will be handled by caller.
            if error is None:
                error = ex
    return error
